File: src/gui/components/ModularComponents.ts
/**
 * 🎨 Modular GUI Components for Luminous Nix
 * TypeScript components that can be reused across different interfaces
 * Integrates with both nixos-gui and Luminous Nix backend
 */

// Component type definitions
export enum ComponentType {
  PROFILE_CARD = 'profile_card',
  PACKAGE_SELECTOR = 'package_selector',
  CONFLICT_RESOLVER = 'conflict_resolver',
  PROGRESS_INDICATOR = 'progress_indicator',
  NATURAL_LANGUAGE_INPUT = 'natural_language_input',
  ESTIMATION_PANEL = 'estimation_panel',
  ROLLBACK_TIMELINE = 'rollback_timeline',
  SYSTEM_PREVIEW = 'system_preview'
}

// Base interfaces
export interface ComponentProps {
  id?: string;
  className?: string;
  style?: Record<string, any>;
}

export interface ComponentEvents {
  onClick?: (event: MouseEvent) => void;
  onChange?: (value: any) => void;
  onSubmit?: (data: any) => void;
}

// Profile types
export interface Profile {
  type: string;
  name: string;
  emoji: string;
  description: string;
  packages: string[];
  optionalPackages?: string[];
  optimizations: string[];
  aiPersona: string;
  settings?: Record<string, any>;
}

export interface ProfileCardProps extends ComponentProps {
  profile: Profile;
  selected?: boolean;
  onSelect?: (profile: Profile) => void;
}

// Package types
export interface Package {
  name: string;
  description?: string;
  version?: string;
  size?: number;
  category?: string;
  selected?: boolean;
  reason?: string;
  dependencies?: string[];
}

export interface PackageSelectorProps extends ComponentProps {
  packages: Package[];
  category: string;
  multiSelect?: boolean;
  showSearch?: boolean;
  showSize?: boolean;
  onPackageSelect?: (packages: Package[]) => void;
}

// Conflict types
export interface Conflict {
  id: string;
  package1: string;
  package2: string;
  reason: string;
  resolution?: string;
  autoResolve?: boolean;
}

export interface ConflictResolverProps extends ComponentProps {
  conflicts: Conflict[];
  onResolve?: (resolutions: Record<string, string>) => void;
  onCancel?: () => void;
}

// Progress types
export interface ProgressIndicatorProps extends ComponentProps {
  total: number;
  current: number;
  message?: string;
  showPercentage?: boolean;
  showETA?: boolean;
  onComplete?: () => void;
}

// Natural language types
export interface NaturalLanguageInputProps extends ComponentProps {
  placeholder?: string;
  suggestions?: string[];
  voiceEnabled?: boolean;
  aiAssist?: boolean;
  onSubmit?: (text: string) => void;
  onVoiceInput?: () => void;
}

// Estimation types
export interface EstimationPanelProps extends ComponentProps {
  packageCount: number;
  estimatedSize?: number;
  estimatedTime?: number;
  showBreakdown?: boolean;
  showMirrorInfo?: boolean;
  onRefresh?: () => void;
}

// Timeline types
export interface Generation {
  id: number;
  description: string;
  date: string;
  packages?: string[];
  changes?: string[];
}

export interface RollbackTimelineProps extends ComponentProps {
  generations: Generation[];
  currentGeneration?: number;
  showDiff?: boolean;
  onSelect?: (generation: Generation) => void;
  onRollback?: (generation: Generation) => void;
}

// System preview types
export interface SystemConfig {
  profile: string;
  packages: string[];
  services: string[];
  settings: Record<string, any>;
}

export interface SystemPreviewProps extends ComponentProps {
  config: SystemConfig;
  showDesktop?: boolean;
  showPackages?: boolean;
  showServices?: boolean;
  interactive?: boolean;
  onEdit?: (section: string) => void;
}

/**
 * Base Component class that all components extend
 */
export abstract class Component<T extends ComponentProps> {
  protected props: T;
  protected element: HTMLElement | null = null;

  constructor(props: T) {
    this.props = props;
  }

  abstract render(): HTMLElement;

  mount(container: HTMLElement): void {
    this.element = this.render();
    container.appendChild(this.element);
  }

  unmount(): void {
    if (this.element && this.element.parentNode) {
      this.element.parentNode.removeChild(this.element);
    }
  }

  update(props: Partial<T>): void {
    this.props = { ...this.props, ...props };
    if (this.element && this.element.parentNode) {
      const parent = this.element.parentNode;
      const newElement = this.render();
      parent.replaceChild(newElement, this.element);
      this.element = newElement;
    }
  }
}

/**
 * Profile Card Component
 */
export class ProfileCard extends Component<ProfileCardProps> {
  render(): HTMLElement {
    const card = document.createElement('div');
    card.className = `profile-card ${this.props.selected ? 'selected' : ''} ${this.props.className || ''}`;
    
    card.innerHTML = `
      <div class="profile-emoji">${this.props.profile.emoji}</div>
      <h3 class="profile-name">${this.props.profile.name}</h3>
      <p class="profile-description">${this.props.profile.description}</p>
      <div class="profile-stats">
        <span class="package-count">📦 ${this.props.profile.packages.length} packages</span>
        <span class="optimizations">⚡ ${this.props.profile.optimizations.join(', ')}</span>
      </div>
      <div class="profile-persona">🤖 Guided by ${this.props.profile.aiPersona}</div>
    `;

    if (this.props.onSelect) {
      card.addEventListener('click', () => {
        this.props.onSelect?.(this.props.profile);
      });
    }

    return card;
  }
}

/**
 * Package Selector Component
 */
export class PackageSelector extends Component<PackageSelectorProps> {
  private selectedPackages: Set<string> = new Set();

  render(): HTMLElement {
    const container = document.createElement('div');
    container.className = `package-selector ${this.props.className || ''}`;

    // Header with category
    const header = document.createElement('h3');
    header.textContent = this.props.category;
    container.appendChild(header);

    // Search bar if enabled
    if (this.props.showSearch) {
      const searchInput = document.createElement('input');
      searchInput.type = 'text';
      searchInput.placeholder = 'Search packages...';
      searchInput.className = 'package-search';
      searchInput.addEventListener('input', (e) => this.filterPackages((e.target as HTMLInputElement).value));
      container.appendChild(searchInput);
    }

    // Package list
    const packageList = document.createElement('div');
    packageList.className = 'package-list';

    this.props.packages.forEach(pkg => {
      const packageItem = this.createPackageItem(pkg);
      packageList.appendChild(packageItem);
    });

    container.appendChild(packageList);

    return container;
  }

  private createPackageItem(pkg: Package): HTMLElement {
    const item = document.createElement('div');
    item.className = 'package-item';
    item.dataset.packageName = pkg.name;

    const checkbox = document.createElement('input');
    checkbox.type = this.props.multiSelect ? 'checkbox' : 'radio';
    checkbox.name = 'packages';
    checkbox.value = pkg.name;
    checkbox.checked = pkg.selected || false;
    checkbox.id = `pkg_${pkg.name}`;

    checkbox.addEventListener('change', () => {
      if (checkbox.checked) {
        this.selectedPackages.add(pkg.name);
      } else {
        this.selectedPackages.delete(pkg.name);
      }
      this.notifySelection();
    });

    const label = document.createElement('label');
    label.htmlFor = `pkg_${pkg.name}`;
    label.innerHTML = `
      <span class="package-name">${pkg.name}</span>
      ${pkg.description ? `<span class="package-description">${pkg.description}</span>` : ''}
      ${pkg.reason ? `<span class="package-reason">${pkg.reason}</span>` : ''}
      ${this.props.showSize && pkg.size ? `<span class="package-size">${this.formatSize(pkg.size)}</span>` : ''}
    `;

    item.appendChild(checkbox);
    item.appendChild(label);

    return item;
  }

  private filterPackages(query: string): void {
    const items = this.element?.querySelectorAll('.package-item') as NodeListOf<HTMLElement>;
    items.forEach(item => {
      const name = item.dataset.packageName?.toLowerCase() || '';
      item.style.display = name.includes(query.toLowerCase()) ? '' : 'none';
    });
  }

  private notifySelection(): void {
    const selected = this.props.packages.filter(p => this.selectedPackages.has(p.name));
    this.props.onPackageSelect?.(selected);
  }

  private formatSize(bytes: number): string {
    const units = ['B', 'KB', 'MB', 'GB'];
    let size = bytes;
    let unitIndex = 0;
    while (size >= 1024 && unitIndex < units.length - 1) {
      size /= 1024;
      unitIndex++;
    }
    return `${size.toFixed(1)} ${units[unitIndex]}`;
  }
}

/**
 * Conflict Resolver Component
 */
export class ConflictResolver extends Component<ConflictResolverProps> {
  private resolutions: Record<string, string> = {};

  render(): HTMLElement {
    const modal = document.createElement('div');
    modal.className = `conflict-resolver-modal ${this.props.className || ''}`;

    const content = document.createElement('div');
    content.className = 'modal-content';

    // Header
    const header = document.createElement('h2');
    header.innerHTML = '⚠️ Package Conflicts Detected';
    content.appendChild(header);

    // Conflicts list
    const conflictsList = document.createElement('div');
    conflictsList.className = 'conflicts-list';

    this.props.conflicts.forEach(conflict => {
      const conflictItem = this.createConflictItem(conflict);
      conflictsList.appendChild(conflictItem);
    });

    content.appendChild(conflictsList);

    // Action buttons
    const actions = document.createElement('div');
    actions.className = 'modal-actions';

    const resolveBtn = document.createElement('button');
    resolveBtn.textContent = 'Resolve Conflicts';
    resolveBtn.className = 'btn-primary';
    resolveBtn.addEventListener('click', () => this.handleResolve());

    const cancelBtn = document.createElement('button');
    cancelBtn.textContent = 'Cancel';
    cancelBtn.className = 'btn-secondary';
    cancelBtn.addEventListener('click', () => this.props.onCancel?.());

    actions.appendChild(resolveBtn);
    actions.appendChild(cancelBtn);
    content.appendChild(actions);

    modal.appendChild(content);
    return modal;
  }

  private createConflictItem(conflict: Conflict): HTMLElement {
    const item = document.createElement('div');
    item.className = 'conflict-item';

    item.innerHTML = `
      <div class="conflict-description">
        <strong>${conflict.package1} ↔ ${conflict.package2}</strong>
        <p>${conflict.reason}</p>
        ${conflict.resolution ? `<p class="suggested-resolution">Suggested: ${conflict.resolution}</p>` : ''}
      </div>
    `;

    const select = document.createElement('select');
    select.className = 'conflict-resolution-select';
    select.innerHTML = `
      <option value="${conflict.package1}">Keep ${conflict.package1}</option>
      <option value="${conflict.package2}">Keep ${conflict.package2}</option>
      <option value="both">Keep both (advanced)</option>
      <option value="neither">Skip both</option>
    `;

    if (conflict.autoResolve && conflict.resolution) {
      select.value = conflict.resolution;
    }

    select.addEventListener('change', () => {
      this.resolutions[conflict.id] = select.value;
    });

    item.appendChild(select);
    return item;
  }

  private handleResolve(): void {
    this.props.onResolve?.(this.resolutions);
  }
}

/**
 * Natural Language Input Component
 */
export class NaturalLanguageInput extends Component<NaturalLanguageInputProps> {
  render(): HTMLElement {
    const container = document.createElement('div');
    container.className = `natural-language-input ${this.props.className || ''}`;

    const inputWrapper = document.createElement('div');
    inputWrapper.className = 'input-wrapper';

    const input = document.createElement('input');
    input.type = 'text';
    input.placeholder = this.props.placeholder || 'Describe what you need...';
    input.className = 'nl-input';

    input.addEventListener('keypress', (e) => {
      if (e.key === 'Enter') {
        this.props.onSubmit?.((e.target as HTMLInputElement).value);
      }
    });

    inputWrapper.appendChild(input);

    // Voice input button if enabled
    if (this.props.voiceEnabled) {
      const voiceBtn = document.createElement('button');
      voiceBtn.className = 'voice-input-btn';
      voiceBtn.innerHTML = '🎤';
      voiceBtn.addEventListener('click', () => this.props.onVoiceInput?.());
      inputWrapper.appendChild(voiceBtn);
    }

    container.appendChild(inputWrapper);

    // Suggestions if provided
    if (this.props.suggestions && this.props.suggestions.length > 0) {
      const suggestionsList = document.createElement('ul');
      suggestionsList.className = 'suggestions';

      this.props.suggestions.forEach(suggestion => {
        const li = document.createElement('li');
        li.className = 'suggestion';
        li.textContent = suggestion;
        li.addEventListener('click', () => {
          input.value = suggestion;
          this.props.onSubmit?.(suggestion);
        });
        suggestionsList.appendChild(li);
      });

      container.appendChild(suggestionsList);
    }

    return container;
  }
}

/**
 * Component Factory for easy creation
 */
export class ComponentFactory {
  static createProfileCard(profile: Profile, selected = false, onSelect?: (p: Profile) => void): ProfileCard {
    return new ProfileCard({ profile, selected, onSelect });
  }

  static createPackageSelector(
    packages: Package[], 
    category: string,
    options: Partial<PackageSelectorProps> = {}
  ): PackageSelector {
    return new PackageSelector({
      packages,
      category,
      multiSelect: true,
      showSearch: true,
      ...options
    });
  }

  static createConflictResolver(
    conflicts: Conflict[],
    onResolve?: (resolutions: Record<string, string>) => void,
    onCancel?: () => void
  ): ConflictResolver {
    return new ConflictResolver({ conflicts, onResolve, onCancel });
  }

  static createNaturalLanguageInput(
    options: Partial<NaturalLanguageInputProps> = {}
  ): NaturalLanguageInput {
    return new NaturalLanguageInput({
      placeholder: 'Describe what you need...',
      voiceEnabled: true,
      aiAssist: true,
      ...options
    });
  }
}

File: src/luminous_nix/gui/modular_components.py
#!/usr/bin/env python3
"""
🎨 Modular GUI Components for Luminous Nix
Reusable components that can be composed for different interfaces
"""

from typing import Dict, Any, List, Optional, Callable
from dataclasses import dataclass
from enum import Enum
import json
from pathlib import Path

# For web integration with existing nixos-gui
import asyncio
import aiohttp
from aiohttp import web


class ComponentType(Enum):
    """Types of reusable components"""
    PROFILE_CARD = "profile_card"
    PACKAGE_SELECTOR = "package_selector"
    CONFLICT_RESOLVER = "conflict_resolver"
    PROGRESS_INDICATOR = "progress_indicator"
    NATURAL_LANGUAGE_INPUT = "natural_language_input"
    ESTIMATION_PANEL = "estimation_panel"
    ROLLBACK_TIMELINE = "rollback_timeline"
    SYSTEM_PREVIEW = "system_preview"


@dataclass
class ComponentSpec:
    """Specification for a modular component"""
    type: ComponentType
    props: Dict[str, Any]
    events: Dict[str, Callable]
    children: List['ComponentSpec'] = None
    
    def to_dict(self) -> Dict:
        """Convert to dictionary for JSON serialization"""
        return {
            'type': self.type.value,
            'props': self.props,
            'children': [c.to_dict() for c in self.children] if self.children else []
        }


class ModularComponentFactory:
    """
    Factory for creating reusable GUI components
    These can be rendered in different contexts (TUI, Web, CLI)
    """
    
    @staticmethod
    def create_profile_card(profile_data: Dict[str, Any], 
                           selected: bool = False) -> ComponentSpec:
        """Create a profile selection card"""
        return ComponentSpec(
            type=ComponentType.PROFILE_CARD,
            props={
                'emoji': profile_data.get('emoji', '📦'),
                'name': profile_data.get('name', 'Unknown'),
                'description': profile_data.get('description', ''),
                'package_count': len(profile_data.get('packages', [])),
                'optimizations': profile_data.get('optimizations', []),
                'selected': selected,
                'style': {
                    'border': 'primary' if selected else 'default',
                    'highlight': selected
                }
            },
            events={
                'onClick': lambda: None,  # Will be overridden by consumer
                'onHover': lambda: None
            }
        )
    
    @staticmethod
    def create_package_selector(packages: List[Dict], 
                               category: str,
                               multi_select: bool = True) -> ComponentSpec:
        """Create a package selection component"""
        return ComponentSpec(
            type=ComponentType.PACKAGE_SELECTOR,
            props={
                'category': category,
                'packages': packages,
                'multi_select': multi_select,
                'show_descriptions': True,
                'show_size': True,
                'grouping': 'category',  # or 'alphabetical', 'popularity'
                'search_enabled': True
            },
            events={
                'onSelect': lambda pkg: None,
                'onDeselect': lambda pkg: None,
                'onSearch': lambda query: None
            }
        )
    
    @staticmethod
    def create_conflict_resolver(conflicts: List[Dict]) -> ComponentSpec:
        """Create a conflict resolution component"""
        return ComponentSpec(
            type=ComponentType.CONFLICT_RESOLVER,
            props={
                'conflicts': conflicts,
                'show_explanations': True,
                'auto_resolve_enabled': True,
                'style': {
                    'severity': 'warning',
                    'modal': True
                }
            },
            events={
                'onResolve': lambda resolution: None,
                'onCancel': lambda: None
            }
        )
    
    @staticmethod
    def create_progress_indicator(total: int, current: int = 0,
                                 message: str = "") -> ComponentSpec:
        """Create a progress indicator"""
        return ComponentSpec(
            type=ComponentType.PROGRESS_INDICATOR,
            props={
                'total': total,
                'current': current,
                'message': message,
                'show_percentage': True,
                'show_eta': True,
                'style': {
                    'color': 'primary',
                    'animated': True
                }
            },
            events={
                'onComplete': lambda: None
            }
        )
    
    @staticmethod
    def create_natural_language_input(placeholder: str = "Describe what you need...",
                                     suggestions: List[str] = None) -> ComponentSpec:
        """Create natural language input component"""
        return ComponentSpec(
            type=ComponentType.NATURAL_LANGUAGE_INPUT,
            props={
                'placeholder': placeholder,
                'suggestions': suggestions or [],
                'show_examples': True,
                'voice_enabled': True,
                'ai_assist': True
            },
            events={
                'onSubmit': lambda text: None,
                'onVoiceInput': lambda: None,
                'onSuggestionSelect': lambda suggestion: None
            }
        )
    
    @staticmethod
    def create_estimation_panel(packages: List[str]) -> ComponentSpec:
        """Create installation estimation panel"""
        return ComponentSpec(
            type=ComponentType.ESTIMATION_PANEL,
            props={
                'package_count': len(packages),
                'estimated_size': len(packages) * 50,  # MB
                'estimated_time': len(packages) * 0.5,  # minutes
                'show_breakdown': True,
                'show_mirror_info': True
            },
            events={
                'onRefresh': lambda: None
            }
        )
    
    @staticmethod
    def create_rollback_timeline(generations: List[Dict]) -> ComponentSpec:
        """Create rollback timeline component"""
        return ComponentSpec(
            type=ComponentType.ROLLBACK_TIMELINE,
            props={
                'generations': generations,
                'show_diff': True,
                'visual_style': 'timeline',  # or 'list', 'cards'
                'preview_enabled': True
            },
            events={
                'onSelect': lambda gen: None,
                'onPreview': lambda gen: None,
                'onRollback': lambda gen: None
            }
        )
    
    @staticmethod
    def create_system_preview(config: Dict) -> ComponentSpec:
        """Create system configuration preview"""
        return ComponentSpec(
            type=ComponentType.SYSTEM_PREVIEW,
            props={
                'config': config,
                'show_desktop': True,
                'show_packages': True,
                'show_services': True,
                'interactive': True
            },
            events={
                'onEdit': lambda section: None,
                'onTest': lambda: None
            }
        )


class ComponentRenderer:
    """
    Base class for rendering components in different contexts
    """
    
    def render(self, component: ComponentSpec) -> Any:
        """Render a component - to be overridden by subclasses"""
        raise NotImplementedError


class WebComponentRenderer(ComponentRenderer):
    """
    Renders components as HTML/JavaScript for web interfaces
    Integrates with existing nixos-gui
    """
    
    def render(self, component: ComponentSpec) -> str:
        """Render component as HTML"""
        renderers = {
            ComponentType.PROFILE_CARD: self._render_profile_card,
            ComponentType.PACKAGE_SELECTOR: self._render_package_selector,
            ComponentType.CONFLICT_RESOLVER: self._render_conflict_modal,
            ComponentType.PROGRESS_INDICATOR: self._render_progress,
            ComponentType.NATURAL_LANGUAGE_INPUT: self._render_nl_input,
            ComponentType.ESTIMATION_PANEL: self._render_estimation,
            ComponentType.ROLLBACK_TIMELINE: self._render_timeline,
            ComponentType.SYSTEM_PREVIEW: self._render_preview
        }
        
        renderer = renderers.get(component.type)
        if renderer:
            return renderer(component)
        return ""
    
    def _render_profile_card(self, component: ComponentSpec) -> str:
        """Render profile card as HTML"""
        props = component.props
        selected_class = 'selected' if props.get('selected') else ''
        
        return f"""
        <div class="profile-card {selected_class}" data-profile="{props.get('name')}">
            <div class="profile-emoji">{props.get('emoji')}</div>
            <h3 class="profile-name">{props.get('name')}</h3>
            <p class="profile-description">{props.get('description')}</p>
            <div class="profile-stats">
                <span class="package-count">📦 {props.get('package_count')} packages</span>
                <span class="optimizations">⚡ {', '.join(props.get('optimizations', []))}</span>
            </div>
        </div>
        """
    
    def _render_package_selector(self, component: ComponentSpec) -> str:
        """Render package selector as HTML"""
        props = component.props
        packages_html = []
        
        for pkg in props.get('packages', []):
            packages_html.append(f"""
                <div class="package-item">
                    <input type="checkbox" 
                           id="pkg_{pkg.get('name')}" 
                           name="packages"
                           value="{pkg.get('name')}"
                           {'checked' if pkg.get('selected') else ''}>
                    <label for="pkg_{pkg.get('name')}">
                        <span class="package-name">{pkg.get('name')}</span>
                        <span class="package-reason">{pkg.get('reason', '')}</span>
                    </label>
                </div>
            """)
        
        return f"""
        <div class="package-selector" data-category="{props.get('category')}">
            <h3>{props.get('category')}</h3>
            {'<input type="text" class="package-search" placeholder="Search packages...">' if props.get('search_enabled') else ''}
            <div class="package-list">
                {''.join(packages_html)}
            </div>
        </div>
        """
    
    def _render_conflict_modal(self, component: ComponentSpec) -> str:
        """Render conflict resolver as modal"""
        props = component.props
        conflicts_html = []
        
        for conflict in props.get('conflicts', []):
            conflicts_html.append(f"""
                <div class="conflict-item">
                    <div class="conflict-description">
                        <strong>{conflict.get('package1')} ↔ {conflict.get('package2')}</strong>
                        <p>{conflict.get('reason')}</p>
                    </div>
                    <div class="conflict-resolution">
                        <select name="resolve_{conflict.get('id')}">
                            <option value="{conflict.get('package1')}">Keep {conflict.get('package1')}</option>
                            <option value="{conflict.get('package2')}">Keep {conflict.get('package2')}</option>
                            <option value="both">Keep both (advanced)</option>
                            <option value="neither">Skip both</option>
                        </select>
                    </div>
                </div>
            """)
        
        return f"""
        <div class="modal conflict-resolver">
            <div class="modal-content">
                <h2>⚠️ Package Conflicts Detected</h2>
                <div class="conflicts-list">
                    {''.join(conflicts_html)}
                </div>
                <div class="modal-actions">
                    <button class="btn-resolve">Resolve Conflicts</button>
                    <button class="btn-cancel">Cancel</button>
                </div>
            </div>
        </div>
        """
    
    def _render_progress(self, component: ComponentSpec) -> str:
        """Render progress indicator"""
        props = component.props
        percentage = (props.get('current', 0) / props.get('total', 1)) * 100
        
        return f"""
        <div class="progress-indicator">
            <div class="progress-message">{props.get('message', '')}</div>
            <div class="progress-bar">
                <div class="progress-fill" style="width: {percentage}%"></div>
            </div>
            <div class="progress-stats">
                <span class="progress-percentage">{percentage:.0f}%</span>
                {'<span class="progress-eta">ETA: calculating...</span>' if props.get('show_eta') else ''}
            </div>
        </div>
        """
    
    def _render_nl_input(self, component: ComponentSpec) -> str:
        """Render natural language input"""
        props = component.props
        
        return f"""
        <div class="natural-language-input">
            <input type="text" 
                   placeholder="{props.get('placeholder')}"
                   class="nl-input">
            {'<button class="voice-input-btn">🎤</button>' if props.get('voice_enabled') else ''}
            {self._render_suggestions(props.get('suggestions', []))}
        </div>
        """
    
    def _render_suggestions(self, suggestions: List[str]) -> str:
        """Render suggestions for natural language input"""
        if not suggestions:
            return ""
        
        items = ''.join([f'<li class="suggestion">{s}</li>' for s in suggestions])
        return f'<ul class="suggestions">{items}</ul>'
    
    def _render_estimation(self, component: ComponentSpec) -> str:
        """Render estimation panel"""
        props = component.props
        
        return f"""
        <div class="estimation-panel">
            <h3>Installation Estimate</h3>
            <div class="estimation-stats">
                <div>📦 {props.get('package_count')} packages</div>
                <div>💾 ~{props.get('estimated_size')} MB</div>
                <div>⏱️ ~{props.get('estimated_time')} minutes</div>
            </div>
            {'<button class="refresh-estimate">Refresh</button>' if props.get('show_breakdown') else ''}
        </div>
        """
    
    def _render_timeline(self, component: ComponentSpec) -> str:
        """Render rollback timeline"""
        props = component.props
        generations = props.get('generations', [])
        
        items = []
        for gen in generations:
            items.append(f"""
                <div class="timeline-item" data-generation="{gen.get('id')}">
                    <div class="timeline-marker"></div>
                    <div class="timeline-content">
                        <h4>Generation {gen.get('id')}</h4>
                        <p>{gen.get('description')}</p>
                        <small>{gen.get('date')}</small>
                    </div>
                </div>
            """)
        
        return f"""
        <div class="rollback-timeline">
            {''.join(items)}
        </div>
        """
    
    def _render_preview(self, component: ComponentSpec) -> str:
        """Render system preview"""
        props = component.props
        
        return f"""
        <div class="system-preview">
            <h3>System Configuration Preview</h3>
            <div class="preview-sections">
                {'<div class="preview-desktop">Desktop Preview</div>' if props.get('show_desktop') else ''}
                {'<div class="preview-packages">Installed Packages</div>' if props.get('show_packages') else ''}
                {'<div class="preview-services">System Services</div>' if props.get('show_services') else ''}
            </div>
        </div>
        """


class ComponentOrchestrator:
    """
    Orchestrates components for different use cases
    Makes it easy to compose interfaces from modular parts
    """
    
    def __init__(self):
        self.factory = ModularComponentFactory()
        self.web_renderer = WebComponentRenderer()
    
    def create_setup_ceremony_components(self, stage: str) -> List[ComponentSpec]:
        """Create components for setup ceremony based on stage"""
        if stage == "welcome":
            return [
                self.factory.create_profile_card(profile, False)
                for profile in self.get_available_profiles()
            ]
        elif stage == "packages":
            return [
                self.factory.create_package_selector(
                    self.get_recommended_packages(),
                    "Essential Packages"
                ),
                self.factory.create_estimation_panel([])
            ]
        elif stage == "conflicts":
            return [
                self.factory.create_conflict_resolver(
                    self.get_current_conflicts()
                )
            ]
        elif stage == "natural":
            return [
                self.factory.create_natural_language_input(
                    "Describe any additional software you need..."
                )
            ]
        elif stage == "preview":
            return [
                self.factory.create_system_preview(
                    self.get_system_config()
                ),
                self.factory.create_progress_indicator(100, 0)
            ]
        return []
    
    def create_dashboard_components(self) -> List[ComponentSpec]:
        """Create components for main dashboard"""
        return [
            self.factory.create_estimation_panel(
                self.get_installed_packages()
            ),
            self.factory.create_rollback_timeline(
                self.get_generations()
            )
        ]
    
    def create_package_manager_components(self) -> List[ComponentSpec]:
        """Create components for package management interface"""
        return [
            self.factory.create_natural_language_input(
                "Search for packages..."
            ),
            self.factory.create_package_selector(
                self.search_packages(""),
                "Available Packages"
            )
        ]
    
    def render_for_web(self, components: List[ComponentSpec]) -> str:
        """Render components for web interface"""
        html_parts = []
        for component in components:
            html_parts.append(self.web_renderer.render(component))
        
        return f"""
        <div class="component-container">
            {''.join(html_parts)}
        </div>
        """
    
    def get_available_profiles(self) -> List[Dict]:
        """Get available user profiles"""
        # This would connect to the ProfileManager
        return [
            {'name': 'Home User', 'emoji': '🏠', 'packages': [], 'optimizations': ['ease']},
            {'name': 'Developer', 'emoji': '⚡', 'packages': [], 'optimizations': ['speed']},
            {'name': 'Creative', 'emoji': '🎨', 'packages': [], 'optimizations': ['gpu']}
        ]
    
    def get_recommended_packages(self) -> List[Dict]:
        """Get recommended packages for current profile"""
        return []
    
    def get_current_conflicts(self) -> List[Dict]:
        """Get current package conflicts"""
        return []
    
    def get_system_config(self) -> Dict:
        """Get current system configuration"""
        return {}
    
    def get_installed_packages(self) -> List[str]:
        """Get list of installed packages"""
        return []
    
    def get_generations(self) -> List[Dict]:
        """Get system generations for rollback"""
        return []
    
    def search_packages(self, query: str) -> List[Dict]:
        """Search for packages"""
        return []


# Web API for integration with nixos-gui
class ComponentWebAPI:
    """
    Web API to serve components to nixos-gui JavaScript frontend
    """
    
    def __init__(self):
        self.orchestrator = ComponentOrchestrator()
        self.app = web.Application()
        self.setup_routes()
    
    def setup_routes(self):
        """Setup web routes"""
        self.app.router.add_get('/api/components/{use_case}', self.get_components)
        self.app.router.add_post('/api/components/render', self.render_components)
    
    async def get_components(self, request):
        """Get components for a specific use case"""
        use_case = request.match_info['use_case']
        stage = request.rel_url.query.get('stage', 'welcome')
        
        if use_case == 'setup_ceremony':
            components = self.orchestrator.create_setup_ceremony_components(stage)
        elif use_case == 'dashboard':
            components = self.orchestrator.create_dashboard_components()
        elif use_case == 'package_manager':
            components = self.orchestrator.create_package_manager_components()
        else:
            components = []
        
        # Convert to JSON-serializable format
        components_data = [c.to_dict() for c in components]
        
        return web.json_response({
            'use_case': use_case,
            'stage': stage,
            'components': components_data
        })
    
    async def render_components(self, request):
        """Render components as HTML"""
        data = await request.json()
        components_data = data.get('components', [])
        
        # Reconstruct ComponentSpec objects
        components = []
        for comp_data in components_data:
            spec = ComponentSpec(
                type=ComponentType(comp_data['type']),
                props=comp_data['props'],
                events={}  # Events are handled client-side
            )
            components.append(spec)
        
        html = self.orchestrator.render_for_web(components)
        
        return web.json_response({
            'html': html
        })
    
    def run(self, host='localhost', port=8888):
        """Run the web API server"""
        web.run_app(self.app, host=host, port=port)


if __name__ == "__main__":
    # Example: Run the component API server
    api = ComponentWebAPI()
    api.run()